fee-monitor/src/DynamicChecker.ts

```diff
@@ -0,0 +1,232 @@
+    }
+    tracerForNextTermSettle.deposit(author, tracerForImmedateSettle.totalFee.minus(distributed));
+}
+
+function setBitIndices(bitset: number[], validatorCount: number): number[] {
+    const indices: number[] = [];
+    for (let i = 0; i < validatorCount; i++) {
+        const arrayIndex = Math.floor(i / 8);
+        const bitIndex = i % 8;
+
+        if ((bitset[arrayIndex] >> bitIndex) & 1) {
+            indices.push(i);
+        }
+    }
+    return indices;
+}
+
+function decodeBitsetField(encodedBitSet: number[]): number[] {
+    const buffer = Buffer.from(encodedBitSet);
+    const decoded = RLP.decode(buffer);
+
+    return Array.from(decoded.values());
+}
+
+export class DynamicChecker {
+    private parentTermValidators: PlatformAddress[] = [];
+    private parentTermId = 0;
+    private parentTermLastBlockNumber = 0;
+    private parentTermFirstBlockNumber = 0;
+    private init = true;
+    private tracerForImmedateSettle = new CCCTracer();
+    private prevTermData: TermData = {
+        intermediateReward: new CCCTracer(),
+        commitedVotesPerValidator: new Map(),
+        assembledVotesPerAuthor: new Map(),
+    };
+    private currentTermData: TermData = {
+        intermediateReward: new CCCTracer(),
+        commitedVotesPerValidator: new Map(),
+        assembledVotesPerAuthor: new Map(),
+    };
+
+    private finalizeTerm(
+        blockNumber: number,
+        termId: number,
+        currentTermValidators: PlatformAddress[],
+    ) {
+        this.prevTermData.intermediateReward = this.currentTermData.intermediateReward;
+        this.prevTermData.assembledVotesPerAuthor = this.currentTermData.assembledVotesPerAuthor;
+        this.prevTermData.commitedVotesPerValidator = this.currentTermData.commitedVotesPerValidator;
+
+        this.currentTermData.intermediateReward = new CCCTracer();
+        this.currentTermData.assembledVotesPerAuthor = new Map();
+        this.currentTermData.commitedVotesPerValidator = new Map();
+
+        this.parentTermFirstBlockNumber = this.parentTermLastBlockNumber + 1;
+        this.parentTermLastBlockNumber = blockNumber;
+        this.parentTermId = termId;
+        this.parentTermValidators = currentTermValidators;
+    }
+
+    private applyAdditionalRewards(validators: PlatformAddress[], totalReducedReward: U64) {
+        const mapValidators: Map<number, PlatformAddress[]> = new Map();
+        for (const validator of validators) {
+            // Wrong assumption?
+            const [missed, proposed] = this.prevTermData.assembledVotesPerAuthor.get(validator) || [
+                0,
+                0,
+            ];
+            if (proposed !== 0) {
+                const ratio = missed / proposed;
+                mapValidators.get(ratio)!.push(validator);
+            }
+        }
+        const sortedValidators = Array.from(mapValidators)
+            .sort(([keyA], [keyB]) => {
+                return keyA === keyB ? 0 : keyA > keyB ? -1 : 1;
```

Comment:
Thank you, I think I can safely assume that the key value is not Infinity or NaN because I only used nonzero value as denominators for the `const ratio`. And used `keyB - keyA` comparing for descending ordering.